pkg/scheduler/framework/plugins/podtopologyspread/plugin.go

```diff
@@ -141,9 +144,122 @@ func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEventWithHint
+			return framework.QueueSkip, nil
+		}
+		// return QueueSkip when the pod labels still match constraints after pod updated.
+		if podLabelsMatchSpreadConstraints(constraints, originalPod.Labels) && podLabelsMatchSpreadConstraints(constraints, modifiedPod.Labels) {
+			return framework.QueueSkip, nil
+		}
+		return framework.Queue, nil
+	}
+
+	// Pod is added. Return Queue when the added Pod has a label that matches with topologySpread's selector.
+	if originalPod != nil {
+		logger.V(5).Info("pod was created", "pod", klog.KObj(pod), "createdPod", klog.KObj(originalPod))
+		if podLabelsMatchSpreadConstraints(constraints, originalPod.Labels) {
+			return framework.Queue, nil
+		}
+		return framework.QueueSkip, nil
+	}
+
+	// Pod is deleted. Return Queue when the deleted Pod has a label that matches with topologySpread's selector.
+	logger.V(5).Info("pod was deleted", "pod", klog.KObj(pod), "deletedPod", klog.KObj(modifiedPod))
+	if podLabelsMatchSpreadConstraints(constraints, modifiedPod.Labels) {
+		return framework.Queue, nil
+	}
+
+	return framework.QueueSkip, nil
+}
+
+func (pl *PodTopologySpread) getConstraints(pod *v1.Pod) ([]topologySpreadConstraint, error) {
+	var constraints []topologySpreadConstraint
+	var err error
+	if len(pod.Spec.TopologySpreadConstraints) > 0 {
+		// We have feature gating in APIServer to strip the spec
+		// so don't need to re-check feature gate, just check length of Constraints.
+		constraints, err = pl.filterTopologySpreadConstraints(
+			pod.Spec.TopologySpreadConstraints,
+			pod.Labels,
+			v1.DoNotSchedule,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("obtaining pod's hard topology spread constraints: %w", err)
+		}
+	} else {
+		constraints, err = pl.buildDefaultConstraints(pod, v1.DoNotSchedule)
+		if err != nil {
+			return nil, fmt.Errorf("setting default hard topology spread constraints: %w", err)
+		}
+	}
+	return constraints, nil
+}
+
+func (pl *PodTopologySpread) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.Queue, err
+	}
+
+	constraints, err := pl.getConstraints(pod)
+	if err != nil {
+		logger.Error(err, "get constraints from pod failed", "pod", klog.KObj(pod))
+		return framework.Queue, err
+	}
+
+	// framework.UpdateNodeLabel: return Queue when topologyKey is added/changed/removed in node's labels, else return QueueSkip.
+	if originalNode != nil && modifiedNode != nil {
+		logger.V(5).Info("node was updated", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		for _, c := range constraints {
+			if originalNode.Labels[c.TopologyKey] != modifiedNode.Labels[c.TopologyKey] {
+				return framework.Queue, nil
+			}
+		}
+		return framework.QueueSkip, nil
```

Comment:
"node was updated" is not enough information to debug, we should be a bit more specific. Can you change other logs as well?

```suggestion
		for _, c := range constraints {
			if originalNode.Labels[c.TopologyKey] != modifiedNode.Labels[c.TopologyKey] {
				logger.V(5).Info("the topology key in the node was updated", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode), "topologyKey", c.TopologyKey)
				return framework.Queue, nil
			}
		}
		logger.V(5).Info("the node's label is updated, but not related to the Pod's topologySpreadConstraints", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
		return framework.QueueSkip, nil
```